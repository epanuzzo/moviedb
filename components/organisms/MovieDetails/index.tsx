import React from "react";
import Link from "next/link";
import Rating from "@/components/atoms/Rating";
import BlockTitle from "@/components/atoms/BlockTitle";
import MovieInfo from "@/components/molecules/MovieInfo";
import ActorDetails, {
  ActorDetailsProps,
} from "@/components/molecules/ActorDetails";
import ReviewDetails, {
  ReviewDetailsProps,
} from "@/components/molecules/ReviewDetails";
import Tag from "@/components/atoms/Tag";
import MovieHeading from "@/components/molecules/MovieHeading";
import { useSelector } from "react-redux";
import { RootState } from "@/redux/store";
import { imageOriginalUrl } from "@/utils/config";

type MovieDetailsProps = {
  id: number;
};

/**
 * Component to display detailed information about a specific movie.
 *
 * @component
 * @param {MovieDetailsProps} props - The properties for the MovieDetails component.
 * @param {number} props.id - The ID of the movie to display details for.
 * @returns {JSX.Element | null} The JSX element for the movie details or null if the movie is not found.
 *
 * @example
 * <MovieDetails id={123} />
 *
 * @remarks
 * This component fetches movie details from the Redux store using the movie ID.
 * It displays various information about the movie including title, tagline, overview,
 * production companies, runtime, genres, cast, and user reviews.
 *
 * @requires
 * - `useSelector` from 'react-redux' to access the Redux store.
 * - `Link` from 'next/link' for navigation.
 * - `Rating`, `MovieHeading`, `MovieInfo`, `Tag`, `BlockTitle`, `ActorDetails`, and `ReviewDetails` components for displaying movie details.
 */
const MovieDetails: React.FC<MovieDetailsProps> = ({ id }) => {
  const { moviesDetails } = useSelector((state: RootState) => state.movie);

  // TODO: Create variables by extracting from movie
  const movie = moviesDetails[id];

  if (!movie) return null;

  return (
    <div className="bg-gray-900 text-white p-6 rounded-lg shadow-lg max-w-4xl mx-auto">
      <div className="mb-6">
        <Link
          href="/"
          className="inline-block bg-gray-700 hover:bg-gray-600 text-white font-semibold py-2 px-4 rounded-md"
        >
          ‚Üê Back to Movies list
        </Link>
      </div>
      <div
        className="relative w-full h-80 bg-cover bg-center rounded-lg overflow-hidden"
        style={{
          backgroundImage: `url(${imageOriginalUrl}${
            movie.backdrop_path || movie.poster_path
          })`,
        }}
      >
        <div className="absolute inset-0 bg-gradient-to-t from-black to-transparent opacity-90"></div>
        <Rating text={movie.vote_average} isLarge />
      </div>

      <MovieHeading
        title={movie.title}
        tagline={movie.tagline}
        description={movie.overview}
      />

      <div className="mt-4 space-y-2">
        {!!movie.production_companies.length && (
          <MovieInfo name="Production">
            {movie.production_companies[0].name}
          </MovieInfo>
        )}
        {movie.runtime && (
          <MovieInfo name="Duration">{`${movie.runtime} minutes`}</MovieInfo>
        )}
        {!!movie.genres.length && (
          <MovieInfo name="Genres">
            {movie.genres.map((genre: { id: number; name: string }) => (
              <Tag key={genre.id}>{genre.name}</Tag>
            ))}
          </MovieInfo>
        )}
      </div>

      {movie.cast && (
        <div className="mt-6">
          <BlockTitle>Cast</BlockTitle>
          <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
            {movie.cast.map((actor: ActorDetailsProps) => (
              <ActorDetails
                key={actor.name}
                name={actor.name}
                role={actor.role}
              />
            ))}
          </div>
        </div>
      )}

      {movie.reviews && (
        <div className="mt-6">
          <BlockTitle>User Reviews</BlockTitle>
          {movie.reviews.map((review: ReviewDetailsProps) => (
            <ReviewDetails key={review.user} {...review} />
          ))}
        </div>
      )}
    </div>
  );
};

export default MovieDetails;
