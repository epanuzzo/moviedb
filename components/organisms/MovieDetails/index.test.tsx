import React from "react";
import renderer from "react-test-renderer";
import MovieDetails from "./index";
import { useDispatch, useSelector } from "react-redux";
import { useRouter } from "next/router";

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock("../../../redux/slices/movieSlice", () => ({
  getMovieDetails: jest.fn(),
}));

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));

describe("MovieDetails", () => {
  let mockDispatch: jest.Mock;
  let mockRouter: unknown;

  beforeEach(() => {
    mockDispatch = jest.fn();
    mockRouter = {
      pathname: "/current-page",
      push: jest.fn(),
      query: { id: "1" },
    };

    (useRouter as unknown as jest.Mock).mockReturnValue(mockRouter);
    (useDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
    (useSelector as unknown as jest.Mock).mockReturnValue({
      movies: [
        { id: 1, title: "Movie 1" },
        { id: 2, title: "Movie 2" },
      ],
      moviesDetails: {
        "1": {
          id: 1,
          title: "Movie 1",
          production_companies: [{ name: "Mock" }],
          genres: [{ id: 1, name: "Action" }],
          cast: [{ name: "Actor 1", role: "Role 1" }],
          reviews: [
            { user: "name", rating: 5, comment: "comment", date: "2021-09-01" },
          ],
        },
      },
      loading: false,
      pages: 2,
      currentPage: 1,
    });
  });

  it("renders the component", () => {
    const tree = renderer.create(<MovieDetails id={1} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders the component with no movie", () => {
    (useSelector as unknown as jest.Mock).mockReturnValueOnce({
      movies: [],
      moviesDetails: {},
      loading: false,
      pages: 0,
      currentPage: 0,
    });
    const tree = renderer.create(<MovieDetails id={1} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
