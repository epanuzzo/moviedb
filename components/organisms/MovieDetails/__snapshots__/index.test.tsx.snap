// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails renders the component 1`] = `
<div
  className="bg-gray-900 text-white p-6 rounded-lg shadow-lg max-w-4xl mx-auto"
>
  <div
    className="mb-6"
  >
    <a
      className="inline-block bg-gray-700 hover:bg-gray-600 text-white font-semibold py-2 px-4 rounded-md"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      ‚Üê Back to Movies list
    </a>
  </div>
  <div
    className="relative w-full h-80 bg-cover bg-center rounded-lg overflow-hidden"
    style={
      {
        "backgroundImage": "url(https://image.tmdb.org/t/p/originalundefined)",
      }
    }
  >
    <div
      className="absolute inset-0 bg-gradient-to-t from-black to-transparent opacity-90"
    />
    <div
      className="absolute top-4 left-4 bg-white text-black rounded-full flex items-center justify-center w-12 h-12"
    >
      <p
        className="text-sm font-bold"
      />
    </div>
  </div>
  <div
    className="mt-6"
  >
    <h1
      className="text-4xl font-bold"
    >
      Movie 1
    </h1>
    <p
      className="text-gray-400 italic"
    />
    <p
      className="mt-4 text-lg"
    />
  </div>
  <div
    className="mt-4 space-y-2"
  >
    <p>
      <span
        className="font-semibold"
      >
        Production
        :
      </span>
       
      Mock
    </p>
    <p>
      <span
        className="font-semibold"
      >
        Genres
        :
      </span>
       
      <span
        className="inline-block bg-gray-800 rounded-full px-3 py-1 text-sm mr-2 my-2"
      >
        Action
      </span>
    </p>
  </div>
  <div
    className="mt-6"
  >
    <h3
      className="text-2xl font-semibold mb-4"
    >
      Cast
    </h3>
    <div
      className="grid grid-cols-1 md:grid-cols-2 gap-4"
    >
      <div
        className="flex items-center space-x-3"
      >
        <div
          className="w-10 h-10 bg-gray-800 rounded-full flex items-center justify-center text-white"
        >
          A
        </div>
        <div>
          <p
            className="font-semibold"
          >
            Actor 1
          </p>
          <p
            className="text-gray-400 text-sm"
          >
            as 
            Role 1
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="mt-6"
  >
    <h3
      className="text-2xl font-semibold mb-4"
    >
      User Reviews
    </h3>
    <div
      className="bg-gray-800 p-4 rounded-lg mb-4"
    >
      <div
        className="flex justify-between"
      >
        <p
          className="font-semibold"
        >
          name
        </p>
        <p
          className="text-sm"
        >
          5
           / 10
        </p>
      </div>
      <p
        className="text-gray-400 mt-2"
      >
        comment
      </p>
      <p
        className="text-xs text-gray-500 mt-2"
      >
        01/09/2021
      </p>
    </div>
  </div>
</div>
`;

exports[`MovieDetails renders the component with no movie 1`] = `null`;
